tiledb/sm/array_schema/dimension.cc

```diff
@@ -1750,8 +1750,8 @@ Status Dimension::check_tile_extent_upper_floor_internal(
   const uint64_t range = domain[1] - domain[0] + 1;
   const T_FLOOR upper_floor =
       ((range - 1) / (tile_extent)) * (tile_extent) + domain[0];
-  const T_FLOOR upper_floor_max =
-      std::numeric_limits<T_FLOOR>::max() - (tile_extent - 1);
+  const T_FLOOR upper_floor_max = std::numeric_limits<T_FLOOR>::max() -
+                                  (static_cast<T_FLOOR>(tile_extent) - 1);
```

Comment:
This eliminates the compiler warning, but doesn't fix the underlying defect that the generates the compiler warning.

The problem arises when `T_FLOOR` is instantiated with a floating point type. The calculations just don't make sense for floating point, since "subtracting one" does nothing when the exponent is large enough, that is, the mantissa doesn't change. 

To make this calculation correct, it needs `enable_if` for selectively instantiate an integral-type version of the function and a floating-point-type version. They can't be correctly defined in the same function, from what I've been able to ascertain.

I came across this defect a year ago. The exact semantics are still somewhat opaque to me. I haven't tackled it myself yet since this is related to some other defects related to tile extents, specifically, the case where the extent is "all floating point numbers", a case where the length of the interval (min-float, max-float) is approximately 2*max-float and thus can't fit in the same floating-point type that the interval is defined with.

Reply:
`T_FLOOR` is always either uint64 or int64. However, `T_EXTENT` can be a floating-point number.

The compiler was converting the int64/uint64 max value to a float to add to the tile extent, then converting the whole thing back to int64/uint64 to store.